web/courses.go

```diff
@@ -533,6 +539,61 @@ func (s *AutograderService) setLastApprovedAssignment(submission *pb.Submission,
 	return s.db.UpdateEnrollment(query)
 }
 
+// acceptRepositoryInvites tries to accept repository invitations for the given repository types on behalf of the given user.
+func (s *AutograderService) acceptRepositoryInvites(ctx context.Context, sc scm.SCM, user *pb.User, course *pb.Course, repoTypes []pb.Repository_Type) {
+	invites := []*github.RepositoryInvitation{}
+	for _, repoType := range repoTypes {
+		repo, err := s.db.GetRepositoryByType(user.ID, course.OrganizationID, repoType)
+		if err != nil {
+			s.logger.Error("Failed to get repo: %s", err)
+			continue
+		}
+		// TODO: pb.Repository should contain Path
+		path := strings.Split(repo.GetHTMLURL(), "/")
+
+		repoInvites, err := sc.GetRepositoryInvites(ctx, &scm.RepositoryOptions{ID: repo.RepositoryID, Owner: course.OrganizationPath, Path: path[len(path)-1]})
+		if err != nil {
+			s.logger.Errorf("Failed to get repository invites for %v: %s - %s", repo.RepoType, user.Login, err)
+			continue
+		}
+		for _, invite := range repoInvites {
+			// Some repositories contain more than one invite
+			// and we only want to accept the invite for the
+			// user we are enrolling.
+			if *invite.Invitee.Login != user.Login {
+				continue
+			}
+			invites = append(invites, invite)
+		}
+	}
+
+	if !(len(invites) > 0) {
+		return
+	}
+
+	user, err := s.db.GetUser(user.ID)
+	if err != nil {
+		s.logger.Errorf("Failed to get user %d: %v", user.ID, err)
+		return
+	}
+	// We need the user access token to accept invitations on their behalf
```

Comment:
We don't need this either. We already pass the `sc` object as input to the `acceptRepositoryInvites` method as well. That is, we only need to query for the access token in one place and we usually pass the `sc` object to those methods that need to talk to GitHub. We typically use this pattern:
```go
	usr, scm, err := s.getUserAndSCMForCourse(ctx, in.GetCourseID())
```
in various methods in `autograder_service.go`.

Reply:
I do not think there is a way around fetching the SCM of the enrollee. Accepting repository invitations has to be done using the access token of the invitee. The `sc` object passed in is that of the teacher, and does not have permission to accept invites for another user.